langgraph/checkpoint/memory.py

```diff
@@ -119,6 +119,57 @@ def list(
                 metadata=self.serde.loads(metadata),
             )
 
+    def search(
+        self,
+        metadata_query: CheckpointMetadata,
+        *,
+        before: Optional[RunnableConfig] = None,
+        limit: Optional[int] = None,
+    ) -> Iterator[CheckpointTuple]:
+        """Search for checkpoints by metadata.
+
+        This method retrieves a list of checkpoint tuples from the in-memory
+        storage based on the provided metadata query. The metadata query does
+        not need to contain all keys defined in the CheckpointMetadata class.
+        The checkpoints are ordered by timestamp in descending order.
+
+        Args:
+            metadata_query (CheckpointMetadata): The metadata query to use for searching the checkpoints.
+            before (Optional[RunnableConfig]): If provided, only checkpoints before the specified timestamp are returned. Defaults to None.
+            limit (Optional[int]): The maximum number of checkpoints to return. Defaults to None.
+
+        Yields:
+            Iterator[CheckpointTuple]: An iterator of checkpoint tuples.
+        """
+        for thread_id, checkpoints in self.storage.items():
+            for ts, (checkpoint_bytes, metadata_bytes) in checkpoints.items():
+                # filter by thread_ts
+                if before and ts >= before["configurable"]["thread_ts"]:
+                    continue
+
+                # check if all query key/value pairs match the metadata
+                metadata = self.serde.loads(metadata_bytes)
+                all_keys_match = all(
+                    query_value == metadata[query_key]
+                    for query_key, query_value in metadata_query.items()
+                )
+
+                # if all query key/value pairs match, yield the checkpoint
+                if all_keys_match:
+                    # limit search results
+                    if limit is not None:
+                        if limit <= 0:
+                            break
+                        limit -= 1
```

Comment:
I tried to copy this logic from `list()`, but the following code raises a `TypeError`:
```python
if limit is not None and limit <= 0:
    break
limit -= 1
```
If `limit` is `None`, then `-=` cannot be used with type `None`. Not sure how this is working in `list()`.

Reply:
let's fix list?

Reply:
Resolved in this PR: https://github.com/langchain-ai/langgraph/pull/447